File: adjacency.py
# adjacency list implementation in python

from collections import defaultdict

def makeList(edges: list):
    d = defaultdict(list)

    for i, j in edges:
        if not len(d[j]):
            d[j] = []
        d[i].append(j)

    return d

def dfs(edges: list):  
    def backtrack(source, target, Dict, count):
        if source == target:
            return 1
        
        for i in Dict[source]:
            count += backtrack(i, target, Dict, 0)
        return count

    d = makeList(edges)

    return backtrack('A', 'D', d, 0)


def bfs(edges: list):  

    source = 'A'
    target = 'D'
    d = makeList(edges)
    count = 0

    q = [source]

    while len(q):
        temp = q.pop()
        
        if temp == target:
            count += 1
            continue
        
        for i in d[temp]:
            q.append(i)
    return count


def main():
    edges = [["A", "B"],  ["B", "C"], ["B", "E"], ["C", "E"], ["E", "D"]]
   
    # num of paths to target node
    print(dfs(edges))
    print(bfs(edges))

    # answer should be 2 with the following list:
    # edges = [["A", "B"],  ["B", "C"], ["B", "E"], ["C", "E"], ["E", "D"]]

if __name__ == '__main__': main()

File: adjacency.cpp
// adjacency list implementation in cpp

#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue> 

typedef std::pair<char, char> charPair;
typedef std::vector<charPair> vChar;
typedef std::unordered_map<char, std::vector<char>> map;

map makeList(vChar v){
	map m;

	for (auto& i: v){
		if (m[i.second].size() == 0){
			m[i.second]  = {};
		}
		m[i.first].push_back(i.second);
	}
	
	return m;
}

int dfs(char source, char target, map m, int count){
	if (source == target) return 1;

	for (auto& i: m[source]){
		count += dfs(i, target, m, 0);
	}
	return count;
}

int bfs(char source, char target, map m){
	int count = 0;
	std::queue<char> q;
	q.push(source);

	while (q.size()){
		char temp = q.front();
		q.pop();

		if (temp == target){
			++count;
			continue;
		}

		for (auto& i: m[source]){
			q.push(i);
		}
	}

	return count;
}


int main(){

	vChar v{
			{'A', 'B'},
			{'B', 'C'},
			{'B', 'E'},
			{'C', 'E'},
			{'E', 'D'}
	};

	auto d = makeList(v);

	// count all paths from A to D
	// result is 2

	std::cout << dfs('A', 'D', d, 0) << std::endl;
	std::cout << bfs('A', 'D', d) << std::endl;

	return 0;
}
